Initial set of RFCs for Az PowerShell (#10045)
Azure/azure-powershell: documentation/RFC/RFC0001-Aure-Profiles.md

---
RFC: RFC0001
Author: Azure PowerShell team
Status:  Draft
SupercededBy: n/a
Version: 1.0
---

# Enable Azure Profiles with Azure PowerShell

Azure comprises several clouds (Public, Azure Stack, Government), each of which comprises several services (APIs) that have different versions.

> Writting an application or a script that is compatible with several cloud can quickly become a very complex task as the developer has to identify for each operation to be performed which API version is compatible on each cloud.

Reviewer: ```suggestion
Writing an application or a script that is compatible with several clouds can quickly become a very complex task as the developer has to identify for each operation to be performed which API version is compatible on each cloud. Since most applications and services in Azure require a set of resources from different services, managing the api-versions of each needed service API in a script quickly multiplies this complexity.
```

> As Azure evolves, existing services will expose new resources or properties and potentially change how to interact with those, subject to breaking changes. We are adding the support of Azure profiles in Azure PowerShell. An Azure profiles is a definition of resources providers and associated API Version.

Reviewer: We are adding support for Azure profiles to help manage this complexity.  Profiles provide a simple version across the entire set of azure services.  When writing a script, simply select a profile for all services, rather than choosing a specific api-version for each service.  This allows scripts to have a consistent api surface that they can be sure will work across the targeted clouds.

Using a profile, a PowerShell script will be attached to a set of known API versions bringing the follwing benefits:

- Compatibility with a specific version of Azure. For example the profile namde `hybrid-2019-03-01`  that will make a script compatible with Azure Stack on version 1904 and above.

> - Stability over time. Your script will be shield from changes happening in the cloud. In order to be able to use the newer capabilities of Azure a different profile would have to be selected.

Reviewer: ```suggestion
- Stability over time. Your script will be shielded from changes happening in the cloud, but still be able to consume the latest bug fixes. However, in order to be able to use the newer capabilities of Azure, a different profile would have to be selected.
```

New profiles will be added on a regular cadence that still needs to be defined and old profiles will be kept available in the latest Azure PowerShell module for some months.

## Motivation

> The support of Azure profiles will support the following scenarios.

Reviewer: ```suggestion
The support of Azure profiles will enable the following scenarios.
```

```code
As a PowerShell script developer
I can attach my script to a profile of Azure

> so that my script does not suffer from future Azure changes

Reviewer: ```suggestion
so that my script does not break when Azure changes
```

```

```code
As a PowerShell script developer
I can select a a profile of Azure

> so that my script will run seamlessly on Azure Stack and Public Azure

Reviewer: ```suggestion
so that my script will run seamlessly on Azure Stack or Azure Sovereign clouds and Public Azure
```
